File: ddolzhenko/rbtree.py
import bst
import networkx as nx
import matplotlib.pyplot as plt

# def is_bst(tree):
#     return bst.is_bst(tree, lambda x: x.is_nil)

from bst import max_node, min_node

def is_bst(tree):
    if tree.is_nil:
        return True

    print(type(tree))
    print(type(tree.left))
    print(type(tree.right))
    print(max_node(tree.left) < tree)

    mn = min_node(tree.right)
    print(mn)
    print(type(mn))
    return  (max_node(tree.left) < tree < min_node(tree.right))  and \
            is_bst(tree.left) and is_bst(tree.right)


def dfs_nodes_inorder(tree):
    if not tree.is_nil:
        yield from dfs_nodes_inorder(tree.left)
        yield tree
        yield from dfs_nodes_inorder(tree.right)



class BaseTree:
    def __init__(self, nil=None):
        self.parent = nil
        self.left =  nil
        self.right = nil
        self._is_black = True
    
    @property
    def is_black(self):
        return self._is_black

    @property
    def is_red(self):
        return not self._is_black

    @property
    def is_nil(self):
        return self == Tree.nil

    def __lt__(self, other):
        return True

    def __str__(self):
        return 'nil'


class Tree(BaseTree):
    nil = BaseTree()

    def __init__(self, x):
        super().__init__(Tree.nil)
        self.data = x
        
    def __lt__(self, other):
        return self.data < other.data

    def __str__(self):
        return str(self.data)

    @property
    def is_valid(self):
        return is_rbtree(self)
    
    def to_graph(self):
        g = nx.Graph()
        for node in dfs_nodes_inorder(self):
            if not node.parent.is_nil:
                g.add_edge(node.data, node.parent.data)
        return g

    def get_positions(node, positions, x, y):
        if not node.is_nil:
            node_size = 0.5
            my_x = (x[1] + x[0]) / 2
            my_y = y[0] + (node_size / 2)
            positions[node.data] = (my_x, -my_y)
            Tree.get_positions(node.left, positions, (x[0], my_x), (my_y, y[1]))
            Tree.get_positions(node.right, positions, (my_x, x[1]), (my_y, y[1]))

    def draw(self):
        positions = {}
        Tree.get_positions(self, positions, x=(0, 10), y=(0, 10))
        g = self.to_graph()

        plt.axis('on')
        nx.draw_networkx(g, positions, node_size=1500, font_size=24, node_color='g')
        plt.show()

def is_rbtree(tree):
    assert is_bst(tree)

    def cond_4_6(tree):
        assert isinstance(tree, BaseTree)

        if tree.is_nil:
            return tree.is_black

        if tree.is_red:
            if tree.right.is_red or tree.left.is_red:
                return False

        return cond_4_6(tree.left) and cond_4_6(tree.right)

    def black_route(tree):
        if tree:
            left_ok, llen = black_route(tree.left)
            right_ok, rlen = black_route(tree.right)
            
            ok = (left_ok and right_ok and llen==rlen)
            return ok, llen + (1 if tree.is_black else 0)

        return True, 0


    return cond_4_6(tree) and black_route(tree)[0]


def insert(root, x):
    assert root.is_valid

    node = bst.insert(root, x)


    assert root.is_valid


def main():

    t = Tree(13)
    t.right = Tree(17)
    t.right.parent = t
    t.right._is_black = False
    
    print(t.data, t.right.data, t.left < t < t.right)

    # exit()


    insert(t, 10)

    t.draw()

if __name__ == '__main__':
    main()

File: ddolzhenko/rb_tree.cpp

template <class T>
struct Node {
    Node(const T data_)
    : left(nullptr)
    , right(nullptr)
    , parent(nullptr)
    , m_is_black(false)
    , data(data_)
    {}


    bool m_is_black;
    Node* left;
    Node* right;
    Node* parent;
    T data;
};

bool is_nil(Node<T>* node)  { return node==nullptr; }

bool is_black(Node<T>* node)  { 
    return is_nil(node) || node->m_is_black; 
}

bool is_red(Node<T>* node)  { return !is_black(node); }

bool is_real(Node<T>* node)  {return !is_nil(node); }

Node<T>* min_node(Node<T>* l, Node<T>* r) {
    if (is_nil(l)) {
        return r;
    }
    if (is_nil(r)) {
        return l;
    }
    return l->data < r->data ? l : r;
}

Node<T>* max_node(Node<T>* l, Node<T>* r) {
    if (is_nil(l)) {
        return r;
    }
    if (is_nil(r)) {
        return l;
    }
    return l->data < r->data ? r : l;
}

Node* max_element(Node* node) {
    if (is_real(node)) {
        node = max_node(node, max_element(node->left));
        node = max_node(node, max_element(node->right));
    }
    return node;
}

Node* min_element(Node* node) {
    if (is_real(node)) {
        node = min_node(node, min_element(node->left));
        node = min_node(node, min_element(node->right));
    }
    return node;
}

bool is_bst(Node* node) {
    if (node->is_nil()) {
        return true;
    }

    auto m = node;
    m = max_node(m, max_element(node->left));
    m = min_node(m, min_element(node->right));
    return m == node && is_bst(node->left) && is_bst(node.right);
}

Node* min_node(Node* node) {
    assert (is_bst(node));

    if (node.is_nil()) {
        return node;
    }

    while (node->left->is_real()) {
        node = node->left;
    }

    return node;
}

Node* max_node(Node* node) {
    assert (is_bst(node));

    if (is_nil(node)) {
        return node;
    }

    while (node->right->is_real()) {
        node = node->right;
    }

    return node;
}


Node* find(Node* node, const T& x) {
    assert (is_bst(node));

    while (is_real(node)) {
        if (node->data < x)
            node = node->right;
        else if (x < node->data)
            node = node->left;
        else
            break;
    }
    return node;
}


Node* insert_bst(Node*& node, Node* what) {
    assert (is_bst(node) &&  is_real(what));

    if (is_nil(node)) {
        // what.parent = node ???
        node = what;
        return node;
    }

    if (what->data < node->data) {
        return insert(node->left, what);
    } else if (node->data < what->data) {
        return insert(node->right, what);
    }
    return node;
}

size_t black_height(Node* node) {
    if (is_nil(node))
        return 0;

    size_t h = is_black(node) ? 1 :0;
    return h + std::max(black_height(node->left), black_height(node->right));
}

bool is_rbtree(Node* node) {
    if (is_nil(node))
        return true;
    
    if (!is_bst(node))
        return false;

    if (is_red(node)) {
        if (!(node->left->is_black() && node.right.is_black())) {
            return false;
        }
    }
    return black_height(node->left) == black_height(node->right);
}


bool is_root(Node* node) {
    return is_nil(node.parent);
}


void fix_insert_rb_tree(Node* node) {
    assert(is_real(node));

    if (is_root(node))
        return;

    auto dad = node.parent;
    if (is_black(dad))
        return;

    if (is_root(dad)) {
        dad->set_black();
    }

    auto grandpa = dad.parent;
    auto uncle = get_uncle(dad);

    if (is_red(uncle)) {
        grandpa.flip();
        dad.flip();
        uncle.flip();
        fix_insert_rb_tree(grandpa);
    }

    if (dad == grandpa.right) {
        auto y = rotate_left(grandpa);
        node->set_black();
        fix_insert_rb_tree(y);
    } else {
        auto y = rotate_right(grandpa);
        node->set_black();
        fix_insert_rb_tree(y);
    }
}

void insert_rbtree(Node*& node, Node* what) {
    assert (is_rbtree(node), && is_real(what));
    auto pos = insert_bst(node, what);
    fix_insert_rb_tree(pos);
}


template <class TKey, class TValue>
class map {
public:
    typedef size_t size_type;
    typedef TKey key_type;
    typedef std::pair<const TKey, TValue> value_type;
    typedef value_type& reference;

    typedef map<TKey, TValue> self_type;

public:
    map() : root(nullptr) {}

    map(const self_type& rhs)
    : root(nullptr) {
        this->assign(rhs.begin(), rhs.end());
    }

    ~map() {
        auto it = begin();
        while (!empty()) {
            it = this->erase(it);
        }
    }


    self_type& operator=(const self_type& rhs) {
        if (this != &rhs) {
            this->swap(self_type(ths));
        }
        return *this;
    }

    vois swap(self_type& other) {
        std::swap(root, other.root);
    }

    template <class TIter>
    void assign(TIter b, TIter e) {
        self_type tmp;
        for (auto it = b; it != e; ++it) {
            tmp.insert(*it);
        }
        swap(tmp);
    }

public:

    iterator begin() {
        return min_node(root);
    }

    iterator end() {
        return nullptr;
    }

    iterator insert(reference x) {
        auto what = new Node<value_type>(x);
        auto where = insert_rbtree(root, what);
        return iterator(where);
    }

    iterator find(const key_type& key) {
        return find(root, key);
    }

    reference operator[](const key_type& key) {
        auto pos = this->find(key);
        if (pos == end()) {
            pos = insert(key, TValue());
        }
        return *pos;
    }

private:
    Node<value_type>* root;
}


